src/org/traccar/database/AttributesManager.java

```diff
@@ -17,183 +17,30 @@
 package org.traccar.database;
 
 import java.sql.SQLException;
-import java.util.Collection;
-import java.util.HashSet;
-import java.util.LinkedList;
-import java.util.Map;
-import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
 
-import org.traccar.Context;
-import org.traccar.helper.Log;
 import org.traccar.model.AttributePermission;
 import org.traccar.model.Attribute;
-import org.traccar.model.Device;
 import org.traccar.model.DeviceAttribute;
 import org.traccar.model.GroupAttribute;
+import org.traccar.model.Identifiable;
 
-public class AttributesManager {
-
-    private final DataManager dataManager;
-
-    private final Map<Long, Attribute> attributes = new ConcurrentHashMap<>();
-    private final Map<Long, Set<Long>> deviceAttributes = new ConcurrentHashMap<>();
-    private final Map<Long, Set<Long>> deviceAttributesWithGroups = new ConcurrentHashMap<>();
-    private final Map<Long, Set<Long>> groupAttributes = new ConcurrentHashMap<>();
-    private final Map<Long, Set<Long>> userAttributes = new ConcurrentHashMap<>();
+public class AttributesManager extends ExtendedObjectManager {
 
     public AttributesManager(DataManager dataManager) {
-        this.dataManager = dataManager;
-        refreshAttributes();
-    }
-
-    public Set<Long> getUserAttributes(long userId) {
-        if (!userAttributes.containsKey(userId)) {
-            userAttributes.put(userId, new HashSet<Long>());
-        }
-        return userAttributes.get(userId);
-    }
-
-    public Set<Long> getGroupAttributes(long groupId) {
-        if (!groupAttributes.containsKey(groupId)) {
-            groupAttributes.put(groupId, new HashSet<Long>());
-        }
-        return groupAttributes.get(groupId);
-    }
-
-    public Set<Long> getDeviceAttributes(long deviceId) {
-        return getDeviceAttributes(deviceAttributes, deviceId);
-    }
-
-    public Set<Long> getAllDeviceAttributes(long deviceId) {
-        return getDeviceAttributes(deviceAttributesWithGroups, deviceId);
+        super(dataManager, Attribute.class, AttributePermission.class, DeviceAttribute.class, GroupAttribute.class);
+        refreshItems();
+        refreshExtendedPermissions();
```

Comment:
Can't we just pass argument with the type that we want to update? If it complicates things probably is not worth it.